### DISCUSSION TITLE
BackendClientError

### DISCUSSION THREAD

**Discussion by yujung7768903** [lablup/contribution-academy-2021#30] [Q&A] [answered]
### Describe the problem
BackendClientError 가 발생합니다.
`Request to the API endpoint has failed` 라는 메세지가 포함돼서 아래 있는 링크를 보고 클라이언트 설정을 해보려고 했지만, 잘 모르겠어서 도움을 요청합니다.
https://client-py.docs.backend.ai/ko/19.03/gsg/config.html

### To Reproduce
dev2103 개발 환경을 설치하고,
client 설치까지 완료한 후에 코드를 돌려보려고 아래의 명령어를 작성했습니다.
`backend.ai run -c 'print("hello world")' python`
그 후에 바로 다음과 같은 에러 메세지가 떴습니다.
```
BackendClientError("Request to the API endpoint has failed.\nCheck your network connection and/or the server status.\n➜ 
ClientConnectorCertificateError(ConnectionKey(host='api.backend.ai', port=443, is_ssl=True, ssl=None, proxy=None, proxy_auth=None, proxy_headers_hash=None), SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1129)'))")
```

### Expected behavior
hello world 가 출력이 되어야합니다.
https://www.youtube.com/watch?v=fgCLzLpaoQ4&t=316s
위 링크를 보며 따라서 실습하였습니다.

### Screenshots
<img width="703" alt="BackendClientError screenshots" src="https://user-images.githubusercontent.com/68562176/128832026-6c4341e2-45b5-4f57-996b-9bd7bf6c26f1.png">

<img width="702" alt="backend ai config" src="https://user-images.githubusercontent.com/68562176/128834501-55f98b92-69cb-4660-a185-3b40c7b770fd.png">

### Environment Description

OS: macOS Big Sur 11.2
App version: 21.03.5
Browser version: Chrome 91.0.4472.114
Python version: 2.7.16
Current Shell: zsh

**Comment by lizable** [comment, marked as answer]
@yujung7768903  Please checkout README in backend.ai-client.
Hint: backend.ai-client-py: [Session mode](https://github.com/lablup/backend.ai-client-py#usage-session-mode)

**Reply by lizable** [reply]
@yujung7768903  Please Follow this guide below.
1. Execute the lines written in [here](https://github.com/lablup/backend.ai/blob/8b20e81169c3056c9dcb193207e61078ba030df9/scripts/install-dev.sh#L671-L682) one-by-one at `../backend.ai-dev/client-py`.
2. Execute `my-backend-ai-session.sh` script
3. Execute `backend.ai --config`.
4. Execute `backend.ai login` and Input ID and PW following [this description](https://github.com/lablup/backend.ai/blob/8b20e81169c3056c9dcb193207e61078ba030df9/scripts/install-dev.sh#L709-L711).

**Reply by yujung7768903** [reply]
> @yujung7768903 Please Follow this guide below.
> 
> 1. Execute the lines written in [here](https://github.com/lablup/backend.ai/blob/8b20e81169c3056c9dcb193207e61078ba030df9/scripts/install-dev.sh#L671-L682) one-by-one at `../backend.ai-dev/client-py`.
> 2. Execute `my-backend-ai-session.sh` script
> 3. Execute `backend.ai --config`.
> 4. Execute `backend.ai login` and Input ID and PW following [this description](https://github.com/lablup/backend.ai/blob/8b20e81169c3056c9dcb193207e61078ba030df9/scripts/install-dev.sh#L709-L711).



#### dev2103 설치후 메세지
<img width="888" alt="스크린샷 2021-08-21 오전 1 16 09" src="https://user-images.githubusercontent.com/68562176/130263159-9dfb154a-6472-4d24-a881-2625cf44876e.png">
* 설치한 폴더(프로젝트 폴더 안)
<img width="719" alt="스크린샷 2021-08-21 오전 1 37 32" src="https://user-images.githubusercontent.com/68562176/130265741-76aff3eb-bb07-41c8-b79c-29cc3dac7e86.png">


#### 발생한 문제
* 가이드에 따라 명령어를 입력한 후 `backend.ai login`을 치고 ID : admin@lablup.com / Password : wJalrXUt 입력했지만 다음과 같이 에러가 발생했습니다.
* `backend.ai config`로 확인해보니 server version이 `failed to fetch`로 나왔습니다.
<img width="711" alt="스크린샷 2021-08-21 오전 1 19 54" src="https://user-images.githubusercontent.com/68562176/130263608-077120bf-6ae5-489d-8354-f61bacf82d4f.png">

* `export BACKEND_ENDPOINT=http://0.0.0.0:${WEBSERVER_PORT}`이 부분에서 webserver_port가 잘못되어서 그런가 싶어 `sudo lsof -PiTCP -sTCP:LISTEN`명령어로 열려있는 포트를 확인해봤습니다.
<img width="658" alt="스크린샷 2021-08-21 오전 1 30 17" src="https://user-images.githubusercontent.com/68562176/130264844-d4a6e310-6fb8-42ac-8274-e104234da6b5.png">

#### 시도한 방법
* 8090은 열려있는 포트에 없길래 혹시 다른 웹서버 포트로 바꿔주면 괜찮아질까 싶어 8080으로 바꿔서 시도해봤습니다.
```
 BACKEND_ENDPOINT=http://0.0.0.0:8080/
```
#### 시도한 결과
다음과 같은 에러가 발생했습니다.
```
aiohttp.client_exceptions.Content Type Error: 0
```
<img width="717" alt="스크린샷 2021-08-21 오전 1 34 49" src="https://user-images.githubusercontent.com/68562176/130265395-03805898-7ac7-48c7-97ca-1c00bffb923e.png">


가이드를 잘못 따라한것인가요? 혹시 해서 dev2103폴더와 client-py폴더에서도 가이드대로 endpoint 설정하고 session으로 실행해봤지만 모두 같은 에러가 발생하였습니다.

**Reply by agatha197** [reply]
I think you didn't run webserver and manager.
For stable execution, run manager, agent, webserver, and storage proxy.
And can you add screenshots of your `my-backend-ai.sh` and `my-backend-ai-session.sh` ? 
When saving secret keys, you wrote `exporeexport`, so I don't know if it's saved well.

**Reply by agatha197** [reply]
Here's the way to run each module.
```
[NOTE] How to run Backend.AI manager:

> cd <manager directory>
> python -m ai.backend.manager.server --debug

[NOTE] How to run Backend.AI agent:

> cd <agent directory>
> python -m ai.backend.agent.server --debug

[NOTE] How to run Backend.AI storage-proxy:

> cd <storage-proxy directory>
> python -m ai.backend.storage.server

[NOTE] How to run Backend.AI web server (for ID/Password login):

> cd <web server directory>
> python -m ai.backend.web.server
```

**Reply by yujung7768903** [reply]
> I think you didn't run webserver and manager.
> For stable execution, run manager, agent, webserver, and storage proxy.
> And can you add screenshots of your `my-backend-ai.sh` and `my-backend-ai-session.sh` ?
> When saving secret keys, you wrote `exporeexport`, so I don't know if it's saved well.

Thank you! Here is screenshots of `my-backend-ai.sh` and `my-backend-ai-session.sh`

* my-backend-ai-session.sh
<img width="711" alt="스크린샷 2021-08-22 오전 12 37 36" src="https://user-images.githubusercontent.com/68562176/130327136-85c4bc4b-ce45-44de-ad95-7f466176ebd3.png">

* my-backend-ai.sh
<img width="628" alt="스크린샷 2021-08-22 오전 12 41 26" src="https://user-images.githubusercontent.com/68562176/130327249-d649f01a-f0c1-40c6-91ff-d7d42a7d90f6.png">

**Reply by agatha197** [reply]
Remove the 5 ~ 8 lines both of them and try again!

**Reply by yujung7768903** [reply]
> Remove the 5 ~ 8 lines both of them and try again!

I didn't run webserver and manager.
After run manager, agent, webserver, and storage proxy, I solved BackendClientError! Thank you👍
